Progress dialog fixes

vadz opened the pull request:
Mostly this contains several important fixes for the native MSW version, but also a couple of changes to the generic version.

Using `AttachThreadInput()` is pretty dangerous, so it would be really great if people could please test this new approach and let me know about any new issues, notably any deadlocks. TIA!

vadz committed:
Remove an apparently unnecessary line from MSW wxProgressDialog

It doesn't seem necessary to disable the use of common buttons, so don't
do it.

vadz committed:
Prevent constant size changes in native MSW wxProgressDialog

MSW implementation of wxProgressDialog adjusted the dialog size to the
size of the message shown in it on each update, resulting in visually
unpleasant constant jumping around (this is the same problem that we
used to have in wxGenericProgressDialog long time ago, see #10624).

Minimize this by using TDM_UPDATE_ELEMENT_TEXT instead of
TDM_SET_ELEMENT_TEXT for changing the element text. This still increases
the dialog size if the new element text is longer than the old value,
but at least doesn't shrink it back if it is shorter, which is already
quite an improvement.

Notice that this change requires using TDF_EXPAND_FOOTER_AREA style, as
otherwise the expanded information can't be updated without a re-layout.
But this doesn't seem to be a big loss and it's not really clear why did
we explicitly clear this flag before anyhow.

Update the dialogs sample to make it easy to test for this behaviour and
the documentation to mention MSW version peculiarities.

vadz committed:
Make wxProgressDialog::Fit() work in native MSW version

This method is supposed to adjust the dialog size to its contents and
while the dialog increases automatically when using native
implementation under MSW, it doesn't shrink back on its own and so it's
still useful to allow Fit() to do it.

Update the sample to test Fit() too.

Commit:
Grow wxGenericProgressDialog automatically if needed

If the new message is longer than the previously shown one, increase the
dialog size to fit it, instead of truncating the message. Still don't do
anything if the new message is shorter to avoid unwanted constant
changes in the dialog size if the message keeps changing.

This is the original patch proposed in #10624, which it really makes
sense to apply now because it makes the generic version match the
behaviour of the native one under MSW (and the behaviour of the native
version cannot be changed).

See #10624.

vadz committed:
Document new wxProgressDialog auto-resizing behaviour

The dialog now grows automatically as needed and Fit() can be called to
shrink it back if desired and this behaviour is consistent under all
platforms.

vadz committed:
Gracefully handle parent disappearance in wxGenericProgressDialog

Use a weak reference to the parent instead of a pointer to avoid
crashing if the parent TLW is destroyed before wxGenericProgressDialog
itself is. This is unlikely to happen in C++ code, where objects of this
class are typically created on the stack, but can happen in e.g.
wxPython where this is not the case.

Closes #16378.

vadz committed:
Allow testing wxGenericProgressDialog in the dialogs sample too

Add the possibility to test the generic implementation of the class when
we use the native one by default, this is useful to allow comparing the
behaviour of the two classes.

vadz committed:
Remove state assigning from wxProgressDialog::GetHandle()

This was added in 01bd848eb9699a60e0bdd5105b9a9c591c9e1b20 without any
explanation and probably was a copy-and-paste typo as it just doesn't
make any sense to change the state of the dialog in this accessor
function (and if the state doesn't change, then this assignment is just
completely useless).

Remove the apparently unnecessary assignment and also an unnecessary
temporary variable.

vadz committed:
Remove state assigning from wxProgressDialog::DoGetPosition() too

This was accidentally copy-and-pasted from another function in
1ef1f8fda6d48594809e11c18194184ba9372b0f, just remove it as it was done
for the original check in the last commit but one.

vadz committed:
Tiny style fix in wxProgressDialog::GetHandle()

No real changes, just remove the trailing spaces and add post-#endif
comment for consistency.

vadz committed:
Return HRESULT, not BOOL, from TaskDialogCallbackProc

Although this doesn't change anything because, by a happy concourse of
circumstances, FALSE is the same as S_OK and TRUE is the same as S_FALSE
numerically, it is still better and more clear, because consistent with
the documentation, to return these constants from the task dialog
callback function rather than boolean ones.

vadz committed:
Make state change in wxProgressDialog::Resume() more explicit

Reset the state to "Continue" instead of assigning it the dialogs own
m_state which was also set to "Continue" from the base class Resume()
called just before, but this wasn't necessarily obvious from just
reading the code.

No real changes.

vadz committed:
Make wxGenericProgressDialog::Resume() virtual

It needs to be overridden in the native MSW wxProgressDialog
implementation, but was only hidden by the function with the same name
in the derived class, making it impossible to call the right function
via the base class pointer, for example.

vadz committed:
Create MSW task dialog progress window with non-null parent

This ensures the correct relationship between wxProgressDialog and its
parent window, fixing different problems with Z-order and the progress
dialog icon, but requires attaching the progress dialog thread input to
the main thread, which creates its own problems: in particular, the task
dialog is now only updated when the messages are dispatched in the main
thread, i.e. more sluggishly than before.

It also requires taking care to avoid the deadlocks as the task dialog
thread now waits for the main thread to dispatch its messages too, as
the child dialog sends messages to the parent window. In particular, we
can't simply wait for the task dialog thread to terminate, but must do
it while dispatching the events as well.

Closes #13185.

vadz committed:
Don't require calling DoNativeBeforeUpdate() with locked CS

Acquire the lock in wxProgressDialog::DoNativeBeforeUpdate() itself
instead of relying on the caller to do it.

This is just a refactoring in preparation for further changes.

vadz committed:
Remove unnecessary check from DoNativeBeforeUpdate()

Don't test if we're using the native task dialog in this function
because it is only called if we are, so simplify the code by omitting
this check.

vadz committed:
Dispatch events from the native MSW wxProgressDialog too

Do this for compatibility with wxGenericProgressDialog, which always did
this and can't really do otherwise as it needs to react to the clicks on
its buttons, and also because not doing it results in the other
application windows being marked as "not responding" by MSW, which looks
bad.

Closes #17768.

vadz committed:
Factor out wxGenericProgressDialog::EnsureActiveEventLoopExists()

Make this function reusable in order to allow it to be called from the
native MSW wxProgressDialog implementation in the upcoming commit.

vadz committed:
Speed up the sample progress dialogs in the sample

Just make them finish faster, it's too boring to wait for them to do it
when testing when they take so long.

vadz committed:
Ensure we don't crash in wxProgressDialog::DoGetPosition()

Check that we do have the shared data before dereferencing the pointer
to it. While this normally will always be the case, it could be null if
some error happened, so add a check for it, just as we already do it
elsewhere.

vadz committed:
Remove unused wxSPDD_DISABLE_ABORT notification

This notification and the code handling it were never used, so just
remove it.

vadz committed:
Factor out wxTopLevelWindow::MSWEnableCloseButton()

Make it possible to reuse this code for other, non-wx, windows.

No real changes, this is just a pure refactoring.

vadz committed:
Disable close title bar button in wxProgressDialog too

When the "Cancel" button inside the dialog is disabled, disable the
close title bar button as well as it serves the same purpose.

In particular, this avoids asserts when clicking the close title bar
button while showing the confirmation message box asking the user
whether the dialog should be cancelled in the dialogs sample.

vadz committed:
Remove redundant wxGenericProgressDialog::m_parentTop assignment

The line setting m_parentTop directly was mistakenly left over even
after 7e083675342b17762735ef2223fb9d264d6fc937 added the call to
SetTopParent() just above it.

Remove it now (better late than never), it's redundant at best and even
wrong if the top level parent can't be used as the dialog parent because
SetTopParent() checks for this but this line did not.

vadz committed:
Just use wxGenericProgressDialog::m_parentTop directly

Remove another redundant line of code from wxGenericProgressDialog::
Create(), there is no need to call GetParentForModalDialog() again when
we just did it to initialize m_parentTop, just use the latter directly.

vadz committed:
Stop using m_winPosition to pass position to the main thread

We can just ask for the window rectangle directly from the main thread,
there is no need to update the position in wxProgressDialogSharedData
all the time.

vadz committed:
Factor out wxProgressDialog::GetTaskDialogRect()

No real changes, just prepare for implementing DoGetSize() in this class
by extracting the common part between the existing DoGetPosition() and
it in a new function.

vadz committed:
Implement wxProgressDialog::DoGetSize() for the native dialog too

DoGetPosition() was done in 1ef1f8fda6d48594809e11c18194184ba9372b0f,
implement DoGetSize() too now in order to give a chance Centre() (which
needs both position and size to work).

See #17768.

vadz committed:
Set wxProgressDialog::m_parent too

This, in addition to the previous commit, allows Centre() to work
correctly as it needs to retrieve the parent window rectangle.

See #17768.

vadz committed:
Avoid spurious assert when cancelling wxProgressDialog

If Cancel/close button was pressed twice in a row, assert checking that
the dialog state was "Continue" could be triggered, which was worse than
annoying as it resulted in a deadlock due to trying to show the assert
dialog box while holding the critical section that prevented the main
thread from continuing.

Allow the state to be either "Continue" or already be set to "Canceled"
now to account for this case.

Still assert for the other invalid states, but they really aren't
supposed to be possible here.

vadz committed:
Avoid deadlock when closing the progress dialog

Don't call EndDialog() while inside the critical section, this could
(and did, sometimes) result in a deadlock if the main thread was trying
to enter it in order to send us wxSPDD_DESTROYED notification as
EndDialog() needs it to process some messages.

vadz committed:
Remove manual wxProgressDialog centering code

This is not necessary any longer now that we use the correct parent for
the dialog window, the task dialog is centered automatically.

And unlike our own code, comctl32.dll code always puts the dialog fully
on one monitor instead of making it span two of them if the parent
window does.

vadz committed:
Use "max" instead of 100 in the progress dialog sample code

No real changes, just don't hardcode 100 as we have a symbolic constant
for it here already.

vadz committed:
Minor optimizations in wxProgressDialog

Don't bother performing the updates if nothing was requested.

And ensure that nothing is requested even more often than it already was
by not requesting an update if the new value is the same as the old one.

vadz commented:
So only a day later I'm not happy with this solution myself any longer :-(

It works sluggishly but acceptably so in the sample which does a lot of updates separated by 100ms sleeps but in a real program which does updates much more rarely (every few seconds), the dialog looks really bad, the progress bar isn't being updated smoothly and the expander area isn't redrawn at all when it's expanded. I don't fully understand why, but calling `AttachThreadInput()` prevents the task dialog from even repainting itself when the event loop in the main thread isn't running (because it waits for the focus/something else update to be acknowledged by the main thread?) and even calling `AttachThreadInput(FALSE)` doesn't help.

I'm not sure what to do, on one hand the new approach is much better from Z-order point of view, all the weird problems we had before are solved by just using the right parent. OTOH it requires pretty frequent updates to the dialog and makes the UI quite unusable otherwise.

a-wi commented:
It seems that position is determined improperly when native dialog is done with progress, e.g. if in MyFrame::ShowProgress() you check the dialog position after the call to DoShowProgress()
`   DoShowProgress(dialog);`
`   wxLogDebug("%i %i", dialog.GetPosition().x, dialog.GetPosition().y);`
you will get the following output:

include\wx/msw/private.h(372): 'GetWindowRect' failed with error 0x00000000 (The operation completed successfully.).
include\wx/msw/private.h(372): 'GetWindowRect' failed with error 0x00000000 (The operation completed successfully.).
-858993460 -858993460

vadz committed:
Don't explicitly attach progress dialog thread input

This is actually completely unnecessary because this is done implicitly
by Windows itself anyhow when we create the dialog with the owner window
belonging to a different thread.

Notice that this also explains why the thread input can't be detached
later.

vadz commented:
@a-wi Yes, the window handle is invalid by then because the dialog had been already destroyed. We probably should reset `m_hwnd` once it happens, but we're not going to get the "correct" position of the non-existing dialog anyhow. I don't think it's a big deal though, why would you query the dialog for position _after_ displaying it?

vadz committed:
Make other wxGenericProgressDialog methods virtual too

Solve the same problem as was recently with hiding, instead of
overriding, Resume() in the native MSW wxProgressDialog class for all
the other methods which were also affected by it.

vadz committed:
Improve progress bar updating in native wxMSW wxProgressDialog

Since the switch to tying the task dialog thread message queue with the
main thread, animating the progress bar didn't work well unless the
dialog was updated very frequently from the main thread and could lag
behind significantly, and confusingly for the user, otherwise.

Work around this by avoiding the progress bar animation and setting it
immediately to its real value. This works much better in practice.

vadz committed:
Get rid of unnecessary variables in wxMSW wxProgressDialog

There doesn't seem to be any need to have both "foo" and "realFoo", just
reuse the existing variables instead.

No real changes.

vadz committed:
Show elapsed/estimated times in wxMSW wxProgressDialog by default

This further improves the dialog usability when the main thread doesn't
update it frequently enough, as these times can be seen immediately
without having to expand the "details" part of the dialog which can be
very sluggish in this case.

It is also more consistent with the generic dialog and the behaviour of
the native dialog before 6b91c5dfab876f0f1b17d54304bfb2fda27398ef which
removed the code clearing TDF_EXPAND_FOOTER_AREA style.

vadz committed:
Remove a now redundant test in UpdateExpandedInformation()

Checking for m_progressBarMarquee is not necessary any longer, just
testing the value is enough.

Update the comments to explain why is it so.

No real changes.

vadz committed:
Check the active event loop in wxGenericProgressDialog dtor

Verify that the active loop didn't change during this object lifetime as
otherwise we could deactivate a different event loop from the one we
installed.

It's still a programming error to write code which doesn't destroy
wxGenericProgressDialog early enough, but at least now "just" assert and
leak memory in this case instead of resetting the active event loop and
hanging the program.

Closes #17983.

vadz committed:
Explain that wxProgressDialog should be created on the stack

Creating it on the heap is unnecessary and can be actually harmful, so
document explicitly that it shouldn't be done.

See #17983.

vadz commented:
I still couldn't find what exactly prevents the task dialog thread from dispatching messages, but with a couple of hacks in 38970e5920 and 21369e84f5, the dialog is much more usable, so I think this should still be merged because now the native dialog is updated about as well (or, rather, badly) as the generic one, but it looks better and doesn't suffer from any Z-order-related problems any more.

I still wish I knew what messages does the dialog thread send to the main thread and why... If only we could somehow prevent them from happening, we could have the absolutely best of both worlds. But as it is, we'll probably have to live with the current version, I can't spend any more time on this.

vadz commented:
I'd really appreciate more testing of this PR, if people using `wxProgressDialog` in their code could please check how it works with this branch and let me know about any (new) problems, it would be great. TIA!

ardovm committed:
Mention possible problem with wxProgressDialog created in OnInit()

Such dialogs must be destroyed before the main event loop starts running
to avoid problems with the temporary event loop created by the dialog
internally.

See #17983.

vadz committed:
Further improve wxProgressDialog resizing behaviour documentation

Unfortunately it doesn't seem possible to prevent the native MSW dialog
from changing its size in both upper and lower direction vertically, so
at least mention this in the documentation and mention a possible
workaround of manually adjusting the text to always have the same number
of lines.

vadz committed:
Get rid of unnecessary wxCriticalSectionLocker use

No real changes, just don't lock a critical section for a short time
only to lock it again almost immediately after unlocking -- just combine
both blocks for which it is locked into one, there is no reason to
release it for TASKDIALOGCONFIG and wxMSWTaskDialogConfig initialization
which are both trivial operations not involving any callbacks.

vadz committed:
Avoid fitting wxProgressDialog to its contents on first update

Use TDM_UPDATE_ELEMENT_TEXT even for the initial update as this allows
to specify the message of roughly comparable (or greater) length than
the messages that will be subsequently used while the dialog is shown
and avoid size changes later, which is much more natural than having to
do it for the first call to Update().

vadz committed:
Split initial wxProgressDialog message consistently with updates

Native wxMSW dialog split the provided message into the title and body
in Update(), but didn't do it for the initial message specified when
constructing the dialog, resulting in weird jumps, due to the font size
change between the body and title dialog elements, if the updated
message differed just slightly from the initial one.

Fix this and refactor the code to reuse the same function for doing this
splitting in both places.

vadz committed:
Add a hint about making wxProgressDialog initially wide enough

Using unbreakable spaces to achieve this looks like a hack, but seems to
be the only thing working well in practice.
